docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,775 @@
+- Verifier
+  - The verifier is responsible for taking in the evidence and endorsements
+    and deciding if the enclave is trustworthy.
+- Relying party
+  - The relying party is the entity interested in communicating with an
+    enclave. The enclave must attest to the relying party before the
+    relying party can trust it. The relying party can also play the role
+    of the verifier, but it does not necessarily have to.
+
+Specification
+-------------
+
+To support custom attestation formats, this document proposes adding a plugin
+model for attestation. The Open Enclave SDK will define a set of common APIs
+that each plugin must implement. Each plugin will define a UUID to distinguish
+it from other plugins.
+
+Futhermore, there will be additional attestation "plugin aware" APIs that are
+analogous to `oe_get_report` and `oe_verify_report` called `oe_get_evidence`
+and `oe_verify_evidence` respectively. These align with the ones proposed in
+the [endorsements design doc](RemoteAttestatonCollaterals.md). There will also
+be functions for registering and unregistering plugins called
+`oe_register_attestation_plugin` and `oe_unregister_attestation_plugin`. The user
+can statically link in their desired plugin and call the register plugin function.
+The attestation data can be retrieved from `oe_get_evidence` with the desired UUID.
+The generated data will have the UUID in its header. The user can `oe_verify_evidence`
+to verify the data and the Open Enclave runtime can use this UUID to determine what plugin
+verification routine to run.
+
+If the plugin is registered on the enclave side, it will only work for the enclave side.
+Likewise, if the plugin is registered for the host side, it will only work for the
+host side. If the user wants to use the plugin for both sides, then they must register
+it once inside the enclave and once inside the host.
+
+### Plugin API
+
+Each plugin must implement the functions below:
+
+```C
+/**
+ * Claims struct used for claims parameters for the plugin.
+ */
+struct oe_claim_t
+{
+    char* name;
+    uint8_t* value;
+    size_t value_size;
+};
+
+/**
+ * Struct that defines the structure of each plugin. Each plugin must
+ * define a UUID for its format and implement the functions in this
+ * struct. Ideally, each plugin should provide a helper function to
+ * create this struct on the behalf of the plugin users.
+ */
+struct oe_attestation_plugin_t
+{
+    /**
+     * The UUID for the plugin.
+     */
+    uuid_t format_id;
+
+    /**
+     * The function that gets executed when a plugin is registered.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] config_data An optional pointer to the configuration data.
+     * @param[in] config_data_size The size in bytes of config_data.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_register)(
+        oe_attestation_plugin_t* plugin_context,
+        const void* config_data,
```

Comment:
Is `config_data` the field that would be used to pass in attestation policy? If so, how would I use different attestation policy per verification? Say, I have two different policies and which one is used is determined based on what operation is requested.

Reply:
I think `oe_verify_evidence` needs parameters for policy. What things are you thinking that would count as the policy? What format would they be in? Could the policies be validated in a plugin agnostic way?

Reply:
I agree that `oe_verify_evidence` should have parameters for policy. @soccerGB earlier design was I believe passing policy through `on_register`, hence my question. Policy is something where caller can express additional assertions regarding the claims in the evidence. This should preferably be in the same format that AAS accepts. In my opinion policies should be plugin agnostic, there should be a common policy verification library shared with AAS. But I think @soccerGB had some other thoughts on this.

Reply:
@jiria, thanks for the reminder.
I had an offline chat with gupta-ak and he will address your feedback.

Reply:
Thanks @soccerGB . @gupta-ak I had a quick look and it does not seem to be yet reflected. Can you please ping me when ready so I can have a look?